### PR TITLE
🍤 Fix Stripe Webhooks

### PR DESCRIPTION
**Pull request by stephencelis**
The fix is fun! The fact we had this bug is not.

### CODE REVIEW

File: Sources/PointFree/Account/Account.swift [hunk 1]
```diff
@@ -596,7 +596,7 @@ private let subscriptionInviteMoreRowView = View<(Stripe.Subscription?, [Databas
 }
 
 private let subscriptionPaymentInfoView = View<Stripe.Subscription> { subscription -> [Node] in
-  guard let card = subscription.customer.sources.data.first else { return [] }
+  guard let card = subscription.customer.right?.sources.data.first else { return [] }
```
Comment (stephencelis) [inline Sources/PointFree/Account/Account.swift]: I've updated all `expand`able properties to be `Either<T.Id, T>`.

File: Sources/PointFree/Account/Change.swift [hunk 2]
```diff
@@ -68,7 +68,7 @@ private func subscriptionChange(_ conn: Conn<StatusLineOpen, (Stripe.Subscriptio
       .flatMap { sub -> EitherIO<Error, Stripe.Subscription> in
         if shouldInvoice {
           parallel(
-            Current.stripe.invoiceCustomer(sub.customer)
+            Current.stripe.invoiceCustomer(sub.customer.either(id, ^\.id))
```
Comment (stephencelis) [inline Sources/PointFree/Account/Change.swift]: It's a nice simple fold to returns an id.

File: Sources/PointFree/SiteMiddleware.swift [hunk 3]
```diff
@@ -201,8 +201,8 @@ private func render(conn: Conn<StatusLineOpen, T3<Database.Subscription?, Databa
 
     case .webhooks(.stripe(.fallthrough)):
       return conn
-        |> writeStatus(.ok)
-        >-> end
+        |> writeStatus(.internalServerError)
+        >-> respond(text: "We don't support this event.")
```
Comment (stephencelis) [inline Sources/PointFree/SiteMiddleware.swift]: Decoding errors fall through here. Now we'll be more aware of failures.

File: Tests/PointFreeTests/StripeHookTests.swift [hunk 4]
```diff
@@ -1,83 +0,0 @@
-import Either
```
Comment (stephencelis) [inline Tests/PointFreeTests/StripeHookTests.swift]: This file moved to `StripeWebhooksTest.swift`. I always try to quick open typing "web" enough to rename.

File: Tests/PointFreeTests/StripeWebhooksTests.swift [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+import WebKit
+#endif
+
+final class StripeWebhooksTests: TestCase {
+  override func setUp() {
+    super.setUp()
+    update(&Current, \.database .~ .mock)
+  }
+
+  func testDecoding() throws {
```
Comment (stephencelis) [inline Tests/PointFreeTests/StripeWebhooksTests.swift]: New test for decoding using simulated payload.

File: Tests/PointFreeTests/__Snapshots__/StripeWebhooksTests/testValidHook.1.Conn.txt [hunk 6]
```diff
@@ -6,7 +6,7 @@
...
+  {"id":"evt_test","data":{"object":{"id":"in_test","amount_paid":1700,"closed":true,"lines":{"has_more":false,"data":[{"amount":1700,"id":"ii_test","quantity":1,"subscription":"sub_test","plan":{"amount":1700,"name":"Individual Monthly","id":"individual-monthly","created":1517356800,"currency":"usd","metadata":{},"interval":"month"}}]},"customer":"cus_test","charge":"ch_test","total":1700,"period_start":1517356800,"amount_remaining":0,"subtotal":1700,"date":1517356800,"period_end":1519948800,"number":"0000000-0000","subscription":"sub_test"}},"type":"invoice.payment_failed"}
 
 ▿ Response
-  Status 200 OK
\ No newline at end of file
+  Status 400 BAD REQUEST
```
Comment (stephencelis) [inline Tests/PointFreeTests/__Snapshots__/StripeWebhooksTests/testValidHook.1.Conn.txt]: Lookin into this...

Comment (stephencelis) [inline Tests/PointFreeTests/__Snapshots__/StripeWebhooksTests/testValidHook.1.Conn.txt]: Fixing these tests now. Need to update the sigs.

File: Sources/PointFree/Account/Change.swift (hunk 2 above)
Comment (mbrandonw) [reply Sources/PointFree/Account/Change.swift]: very nice

File: Tests/PointFreeTests/__Snapshots__/StripeWebhooksTests/testValidHook.1.Conn.txt (hunk 6 above)
Comment (mbrandonw) [reply Tests/PointFreeTests/__Snapshots__/StripeWebhooksTests/testValidHook.1.Conn.txt]: so we expect a 400 on this?

File: Sources/PointFree/Stripe.swift [hunk 7]
```diff
@@ -128,7 +128,7 @@ public struct Stripe {
   public struct Invoice: Codable {
     public private(set) var amountDue: Cents
     public private(set) var amountPaid: Cents
-    public private(set) var charge: Charge?
+    public private(set) var charge: Either<Charge.Id, Charge>?
```
Comment (stephencelis) [inline Sources/PointFree/Stripe.swift]: This is the root of the bug. When we added support to view historic invoices, we needed to expand information to see the charge. Stripe webhooks do _not_ expand this kind of info, so we need to support the unexpanded state. `Either` and conditional conformance makes this simple!

**Pull request merged**